core/arch/aarchxx/mangle.c

```diff
@@ -219,6 +219,99 @@ insert_clear_eflags(dcontext_t *dcontext, clean_call_info_t *cci,
      */
 }
 
+#ifdef AARCH64
+/* Creates a memory reference for registers saved/restored to memory. */
+static opnd_t
+create_base_disp_for_save_restore(uint base_reg, uint first_reg, uint reg,
+                                  bool is_single_reg, bool is_gpr)
+{
+
+    uint opsz;
+    if (is_gpr)
+        if (is_single_reg)
+            opsz = OPSZ_8;
+        else
+            opsz = OPSZ_16;
+    else
+        if (is_single_reg)
+            opsz = OPSZ_16;
+        else
+            opsz = OPSZ_32;
+
+    uint offset = is_gpr ? REG_OFFSET(DR_REG_X0 + reg) : reg * sizeof(dr_simd_t);
+    return opnd_create_base_disp(base_reg, DR_REG_NULL, 0, offset, opsz);
+}
+
+/* Creates code to save or restore GPR or SIMD registers to memory starting at
+ * base_reg. Uses stp/ldp to save/restore as many register pairs to memory as possible
+ * and uses a single str/ldp for the last register in case the number of registers
+ * is uneven. Optionally takes reg_skip into account.
+ */
+static void
+insert_save_or_restore_registers(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+                                 bool *reg_skip, reg_id_t base_reg, reg_id_t first_reg,
+                                 bool save, bool is_gpr)
+{
+    uint i, reg1 = UINT_MAX, num_regs = is_gpr ? 30 : 32;
+    instr_t *new_instr;
+    /* Use stp/ldp to save/restore as many register pairs to memory, skipping
+     * registers according to reg_skip.
+     */
+    for (i = 0; i < num_regs; i += 1) {
+        if (reg_skip != NULL && reg_skip[i])
+            continue;
+
+        if (reg1 == UINT_MAX)
+            reg1 = i;
+        else {
+            opnd_t mem = create_base_disp_for_save_restore(base_reg, first_reg, reg1,
+                                                           false /* is_single_reg */,
+                                                           is_gpr);
+            if (save) {
+                new_instr = INSTR_CREATE_stp(dcontext, mem,
+                                             opnd_create_reg(first_reg + reg1),
+                                             opnd_create_reg(first_reg + i));
+            } else {
+                new_instr = INSTR_CREATE_ldp(dcontext, opnd_create_reg(first_reg + reg1),
+                                             opnd_create_reg(first_reg + i), mem);
+            }
+            PRE(ilist, instr, new_instr);
+            reg1 = UINT_MAX;
+        }
+    }
+
+    /* Use str/ldr to save/restore last single register to memory if the number
+     * of registers to save/restore is uneven.
```

Comment:
nit: I am not native speaker, does "uneven" mean "odd", I mostly see not even or odd.
http://www.thefreedictionary.com/uneven
It seems it could mean odd.